lisa/transformer.py

```diff
@@ -0,0 +1,168 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import copy
+import functools
+from typing import Any, Dict, List, Set
+
+from lisa import schema
+from lisa.parameter_parser.runbook import RunbookBuilder
+from lisa.util import LisaException, constants, subclasses
+from lisa.util.logger import get_logger
+from lisa.variable import VariableEntry, merge_variables, replace_variables
+
+_get_init_logger = functools.partial(get_logger, "init", "transformer")
+
+
+class Transformer(subclasses.BaseClassWithRunbookMixin):
+    def __init__(self, runbook: schema.Transformer, *args: Any, **kwargs: Any) -> None:
+        super().__init__(runbook, *args, **kwargs)
+        self.name = runbook.name
+        self.prefix = runbook.prefix
+        self._log = get_logger("transformer", self.name)
+
+        self.depend_on = runbook.depend_on
+        self.rename = runbook.rename
+
+    def run(self, is_dry_run: bool = False) -> Dict[str, VariableEntry]:
+        if is_dry_run:
+            # create mock up variables and validate
+            output_names = self._output_names
+            # add prefix
+            variables = {x: VariableEntry(x, "mock value") for x in output_names}
+        else:
+            self._log.info("transformer is running.")
+            variables = self._internal_run()
+
+        results: Dict[str, VariableEntry] = dict()
+        unmatched_renames = copy.copy(self.rename)
+        for name, value in variables.items():
+            name = f"{self.prefix}_{name}"
+            if name in self.rename:
+                del unmatched_renames[name]
+                name = self.rename[name]
+            results[name] = VariableEntry(name, value)
+        if unmatched_renames:
+            raise LisaException(f"unmatched rename variable: {unmatched_renames}")
+        return results
+
+    @property
+    def _output_names(self) -> List[str]:
+        """
+        List names of outputs, which are returned after run.
+        """
+        raise NotImplementedError()
+
+    def _internal_run(self) -> Dict[str, Any]:
+        raise NotImplementedError()
+
+
+def _sort(transformers: List[schema.Transformer]) -> List[schema.Transformer]:
```

Comment:
This topological sort could have been done with less code (and more performant code), by e.g. using deques. Your call.

Reply:
Do you mean the [deque](https://towardsdatascience.com/python-deques-learn-what-deques-are-designed-for-7b345a7adb80#:~:text=Key%20Takeaways%20and%20Conclusion%201%20Deque%20is%20short,access%20to%20the%20elements%20in%20the%20container%20object.)? It looks a double-end queue, not solving the sort problem.

Reply:
Yes, deque. Of course it does not solve the problem, but there are better algorithms--using that data structure--to accomplish the task with less time complexity (e.g. https://guides.codepath.com/compsci/Topological-Sort)

Reply:
We can also use a library called graphlib. Example is given here : https://www.geeksforgeeks.org/topological-sorting-using-graphlib-python-module/

Reply:
Thank you all. I will implement it by topological sorting. The graphlib is introduced from 3.9, I don't want to spend time to upgrade Python on each environments right now.